Found non-deterministic tests.

**Describe the bug**
The following tests fail due to an assertCheck between two JSON Strings.

1. [testOptionalSerialization()](https://github.com/EliasStar/JsonRpc4J/blob/6d7f58389e86efa4f44cc8906a7d1963b7f208b9/src/test/java/eliasstar/gson/OptionalTypeAdapterTests.java#L43)
2. [testNotificationSending()](https://github.com/EliasStar/JsonRpc4J/blob/6d7f58389e86efa4f44cc8906a7d1963b7f208b9/src/test/java/eliasstar/jsonrpc/ConnectionTests.java#L153)
3. [testNotificationSerialization()](https://github.com/EliasStar/JsonRpc4J/blob/6d7f58389e86efa4f44cc8906a7d1963b7f208b9/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java#L42)
4. [testRequestIdSerialization()](https://github.com/EliasStar/JsonRpc4J/blob/6d7f58389e86efa4f44cc8906a7d1963b7f208b9/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java#L49)
5. [testRequestParamsSerialization()](https://github.com/EliasStar/JsonRpc4J/blob/6d7f58389e86efa4f44cc8906a7d1963b7f208b9/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java#L73)

The primary issue in these functions was the reliance on the specific ordering of keys in the serialized JSON strings for comparisons in the test assertions. In JSON, the order of keys is not guaranteed and can vary depending on how the JSON serialization is implemented. This variability led to test failures, even though the JSON strings were structurally equivalent and contained the same data.

**Fix:**
The fix involved using Gson, a popular JSON processing library in Java, to parse the expected and actual JSON strings into JsonObject instances before comparing them. By doing so, the comparison became order-independent, focusing solely on the structure and content of the JSON data. This approach ensured that the tests were robust against differences in key ordering, thereby accurately validating the correctness of the JSON serialization logic regardless of the order in which the keys appeared in the serialized JSON strings. The fix was applied consistently across all functions, streamlining the test verification process and improving the reliability of the tests.

**How to reproduce**

**Steps to reproduce the behavior:**
I used an open-source tool called [NonDex](https://github.com/TestingResearchIllinois/NonDex) to detect the assumption by shuffling the order of returned exception types.
Running the following commands will test the aforementioned operation

**Clone the Repo**
```
https://github.com/EliasStar/JsonRpc4J.git
```
**Compile the project**
```
mvn install -am -DskipTests
```
**(Optional) Run the unit test**
```
mvn test
```
**Run the unit test using NonDex**
```
mvn edu.illinois:nondex-maven-plugin:2.1.1:nondex
```

Stack trace for additional information:
```
[ERROR] Failures: 
[ERROR]   OptionalTypeAdapterTests.testOptionalSerialization:50 expected: <{"testString":null,"testOptional":"test"}> but was: <{"testOptional":"test","testString":null}>
[ERROR]   ConnectionTests.testNotificationSending:171 Multiple Failures (2 failures)
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","method":"test","params":[]}> but was: <{"params":[],"jsonrpc":"2.0","method":"test"}>
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","method":"test","params":{}}> but was: <{"params":{},"jsonrpc":"2.0","method":"test"}>
[ERROR]   NotificationTests.testNotificationSerialization:46 Multiple Failures (2 failures)
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","method":"test"}> but was: <{"method":"test","jsonrpc":"2.0"}>
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","method":"test"}> but was: <{"method":"test","jsonrpc":"2.0"}>
[ERROR]   RequestTests.testRequestIdSerialization:56->lambda$testRequestIdSerialization$2:56 Multiple Failures (2 failures)
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","id":"test","method":"test"}> but was: <{"jsonrpc":"2.0","method":"test","id":"test"}>
        org.opentest4j.AssertionFailedError: expected: <{"jsonrpc":"2.0","id":"test","method":"test"}> but was: <{"method":"test","id":"test","jsonrpc":"2.0"}>
[ERROR]   RequestTests.testRequestParamsSerialization:79 expected: <{"jsonrpc":"2.0","method":"test","params":["test",0,null]}> but was: <{"params":["test",0,null],"method":"test","jsonrpc":"2.0"}>

```

--- a/src/test/java/eliasstar/gson/OptionalTypeAdapterTests.java
+++ b/src/test/java/eliasstar/gson/OptionalTypeAdapterTests.java
@@ -24,8 +24,8 @@ import java.util.Optional;
 
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
-
 import com.google.gson.JsonObject;
+
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
@@ -48,11 +48,11 @@ public final class OptionalTypeAdapterTests {
 
         testObj.testOptional = Optional.of("test");
         assertEquals("{\"testOptional\":\"test\"}", gson.toJson(testObj));
-        assertEquals(gson.fromJson("{\"testString\":null,\"testOptional\":\"test\"}", JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
+        assertEquals(gsonWithNulls.fromJson("{\"testString\":null,\"testOptional\":\"test\"}", JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
 
         testObj.testString = "test";
         assertEquals(gson.fromJson("{\"testString\":\"test\",\"testOptional\":\"test\"}", JsonObject.class), gson.fromJson(gson.toJson(testObj), JsonObject.class));
-        assertEquals(gson.fromJson("{\"testString\":\"test\",\"testOptional\":\"test\"}",JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
+        assertEquals(gsonWithNulls.fromJson("{\"testString\":\"test\",\"testOptional\":\"test\"}", JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
 
         testObj.testOptional = Optional.empty();
         assertEquals("{\"testString\":\"test\"}", gson.toJson(testObj));

--- a/src/test/java/eliasstar/gson/OptionalTypeAdapterTests.java
+++ b/src/test/java/eliasstar/gson/OptionalTypeAdapterTests.java
@@ -25,6 +25,7 @@ import java.util.Optional;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 
+import com.google.gson.JsonObject;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
@@ -47,11 +48,11 @@ public final class OptionalTypeAdapterTests {
 
         testObj.testOptional = Optional.of("test");
         assertEquals("{\"testOptional\":\"test\"}", gson.toJson(testObj));
-        assertEquals("{\"testString\":null,\"testOptional\":\"test\"}", gsonWithNulls.toJson(testObj));
+        assertEquals(gson.fromJson("{\"testString\":null,\"testOptional\":\"test\"}", JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
 
         testObj.testString = "test";
-        assertEquals("{\"testString\":\"test\",\"testOptional\":\"test\"}", gson.toJson(testObj));
-        assertEquals("{\"testString\":\"test\",\"testOptional\":\"test\"}", gsonWithNulls.toJson(testObj));
+        assertEquals(gson.fromJson("{\"testString\":\"test\",\"testOptional\":\"test\"}", JsonObject.class), gson.fromJson(gson.toJson(testObj), JsonObject.class));
+        assertEquals(gson.fromJson("{\"testString\":\"test\",\"testOptional\":\"test\"}",JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(testObj), JsonObject.class));
 
         testObj.testOptional = Optional.empty();
         assertEquals("{\"testString\":\"test\"}", gson.toJson(testObj));

--- a/src/test/java/eliasstar/jsonrpc/ConnectionTests.java
+++ b/src/test/java/eliasstar/jsonrpc/ConnectionTests.java
@@ -156,15 +156,21 @@ public final class ConnectionTests {
         var cases = new Executable[] {
                 () -> {
                     connection.sendNotification("test");
-                    assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\"}", client.getRequest());
+                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\"}", JsonObject.class);
+                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
+                    assertEquals(expected, actual);
                 },
                 () -> {
                     connection.sendNotification("test", new JsonArray());
-                    assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[]}", client.getRequest());
+                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[]}", JsonObject.class);
+                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
+                    assertEquals(expected, actual);
                 },
                 () -> {
                     connection.sendNotification("test", new JsonObject());
-                    assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{}}", client.getRequest());
+                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{}}", JsonObject.class);
+                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
+                    assertEquals(expected, actual);
                 }
         };
 

--- a/src/test/java/eliasstar/jsonrpc/ConnectionTests.java
+++ b/src/test/java/eliasstar/jsonrpc/ConnectionTests.java
@@ -156,21 +156,18 @@ public final class ConnectionTests {
         var cases = new Executable[] {
                 () -> {
                     connection.sendNotification("test");
-                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\"}", JsonObject.class);
-                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
-                    assertEquals(expected, actual);
+                    var json = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\"}", JsonObject.class);
+                    assertEquals(json, gson.fromJson(client.getRequest(), JsonObject.class));
                 },
                 () -> {
                     connection.sendNotification("test", new JsonArray());
-                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[]}", JsonObject.class);
-                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
-                    assertEquals(expected, actual);
+                    var json = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[]}", JsonObject.class);
+                    assertEquals(json, gson.fromJson(client.getRequest(), JsonObject.class));
                 },
                 () -> {
                     connection.sendNotification("test", new JsonObject());
-                    JsonObject expected = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{}}", JsonObject.class);
-                    JsonObject actual = gson.fromJson(client.getRequest(), JsonObject.class);
-                    assertEquals(expected, actual);
+                    var json = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{}}", JsonObject.class);
+                    assertEquals(json, gson.fromJson(client.getRequest(), JsonObject.class));
                 }
         };
 

--- a/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
@@ -47,8 +47,7 @@ public final class NotificationTests {
         JsonObject expectedJson = gson.fromJson(json, JsonObject.class);
         assertAll(
                 () -> assertEquals(expectedJson, gson.fromJson(gson.toJson(obj), JsonObject.class)),
-                () -> assertEquals(expectedJson, gsonWithNulls.fromJson(gsonWithNulls.toJson(obj), JsonObject.class))
-        );
+                () -> assertEquals(expectedJson, gsonWithNulls.fromJson(gsonWithNulls.toJson(obj), JsonObject.class)));
     }
 
     @Test

--- a/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
@@ -22,6 +22,7 @@ import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import com.google.gson.Gson;
+import com.google.gson.JsonObject;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
@@ -43,9 +44,11 @@ public final class NotificationTests {
         var json = "{\"jsonrpc\":\"2.0\",\"method\":\"test\"}";
         var obj = new Notification("test");
 
+        JsonObject expectedJson = gson.fromJson(json, JsonObject.class);
         assertAll(
-                () -> assertEquals(json, gson.toJson(obj)),
-                () -> assertEquals(json, gsonWithNulls.toJson(obj)));
+                () -> assertEquals(expectedJson, gson.fromJson(gson.toJson(obj), JsonObject.class)),
+                () -> assertEquals(expectedJson, gsonWithNulls.fromJson(gsonWithNulls.toJson(obj), JsonObject.class))
+        );
     }
 
     @Test

--- a/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/NotificationTests.java
@@ -41,13 +41,12 @@ public final class NotificationTests {
 
     @Test
     public void testNotificationSerialization() {
-        var json = "{\"jsonrpc\":\"2.0\",\"method\":\"test\"}";
+        var json = gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\"}", JsonObject.class);
         var obj = new Notification("test");
 
-        JsonObject expectedJson = gson.fromJson(json, JsonObject.class);
         assertAll(
-                () -> assertEquals(expectedJson, gson.fromJson(gson.toJson(obj), JsonObject.class)),
-                () -> assertEquals(expectedJson, gsonWithNulls.fromJson(gsonWithNulls.toJson(obj), JsonObject.class)));
+                () -> assertEquals(json, gson.fromJson(gson.toJson(obj), JsonObject.class)),
+                () -> assertEquals(json, gsonWithNulls.fromJson(gsonWithNulls.toJson(obj), JsonObject.class)));
     }
 
     @Test

--- a/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
@@ -24,6 +24,7 @@ import static org.junit.jupiter.api.Assertions.assertEquals;
 import java.util.HashMap;
 
 import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonNull;
 import com.google.gson.JsonObject;
@@ -54,8 +55,8 @@ public final class RequestTests {
         cases.put("{\"jsonrpc\":\"2.0\",\"id\":null,\"method\":\"test\"}", new Request(NullId.instance(), "test", null));
 
         cases.forEach((e, a) -> assertAll(
-                () -> assertEquals(e, gson.toJson(a)),
-                () -> assertEquals(e, gsonWithNulls.toJson(a))));
+                () -> assertEquals(gson.fromJson(e, JsonObject.class), gson.fromJson(gson.toJson(a), JsonObject.class)),
+                () -> assertEquals(gsonWithNulls.fromJson(e, JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(a), JsonObject.class))));
     }
 
     @Test
@@ -71,20 +72,23 @@ public final class RequestTests {
 
     @Test
     public void testRequestParamsSerialization() {
-        var arrParams = new JsonArray();
+        // Test array parameters
+        JsonArray arrParams = new JsonArray();
         arrParams.add("test");
         arrParams.add(0);
         arrParams.add(JsonNull.INSTANCE);
+        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[\"test\",0,null]}", JsonObject.class),
+                gson.fromJson(gson.toJson(new Notification("test", arrParams)), JsonObject.class));
 
-        assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[\"test\",0,null]}", gson.toJson(new Notification("test", arrParams)));
-
-        var objParams = new JsonObject();
+        // Test object parameters
+        JsonObject objParams = new JsonObject();
         objParams.addProperty("string", "test");
         objParams.addProperty("number", 0);
         objParams.add("null", JsonNull.INSTANCE);
-
-        assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0}}", gson.toJson(new Notification("test", objParams)));
-        assertEquals("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0,\"null\":null}}", gsonWithNulls.toJson(new Notification("test", objParams)));
+        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0}}", JsonObject.class),
+                gson.fromJson(gson.toJson(new Notification("test", objParams)), JsonObject.class));
+        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0,\"null\":null}}", JsonObject.class),
+                gsonWithNulls.fromJson(gsonWithNulls.toJson(new Notification("test", objParams)), JsonObject.class));
     }
 
     @Test

--- a/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
@@ -72,7 +72,6 @@ public final class RequestTests {
 
     @Test
     public void testRequestParamsSerialization() {
-        // Test array parameters
         JsonArray arrParams = new JsonArray();
         arrParams.add("test");
         arrParams.add(0);
@@ -80,7 +79,6 @@ public final class RequestTests {
         assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[\"test\",0,null]}", JsonObject.class),
                 gson.fromJson(gson.toJson(new Notification("test", arrParams)), JsonObject.class));
 
-        // Test object parameters
         JsonObject objParams = new JsonObject();
         objParams.addProperty("string", "test");
         objParams.addProperty("number", 0);

--- a/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
+++ b/src/test/java/eliasstar/jsonrpc/objects/RequestTests.java
@@ -24,7 +24,6 @@ import static org.junit.jupiter.api.Assertions.assertEquals;
 import java.util.HashMap;
 
 import com.google.gson.Gson;
-import com.google.gson.GsonBuilder;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonNull;
 import com.google.gson.JsonObject;
@@ -72,21 +71,26 @@ public final class RequestTests {
 
     @Test
     public void testRequestParamsSerialization() {
-        JsonArray arrParams = new JsonArray();
+        var arrParams = new JsonArray();
         arrParams.add("test");
         arrParams.add(0);
         arrParams.add(JsonNull.INSTANCE);
-        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[\"test\",0,null]}", JsonObject.class),
-                gson.fromJson(gson.toJson(new Notification("test", arrParams)), JsonObject.class));
 
-        JsonObject objParams = new JsonObject();
+        var arrNotification = new Notification("test", arrParams);
+        var arrJson = "{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":[\"test\",0,null]}";
+        assertEquals(gson.fromJson(arrJson, JsonObject.class), gson.fromJson(gson.toJson(arrNotification), JsonObject.class));
+
+        var objParams = new JsonObject();
         objParams.addProperty("string", "test");
         objParams.addProperty("number", 0);
         objParams.add("null", JsonNull.INSTANCE);
-        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0}}", JsonObject.class),
-                gson.fromJson(gson.toJson(new Notification("test", objParams)), JsonObject.class));
-        assertEquals(gson.fromJson("{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0,\"null\":null}}", JsonObject.class),
-                gsonWithNulls.fromJson(gsonWithNulls.toJson(new Notification("test", objParams)), JsonObject.class));
+
+        var objNotification = new Notification("test", objParams);
+        var objJson = "{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0}}";
+        var objJsonNull = "{\"jsonrpc\":\"2.0\",\"method\":\"test\",\"params\":{\"string\":\"test\",\"number\":0,\"null\":null}}";
+
+        assertEquals(gson.fromJson(objJson, JsonObject.class), gson.fromJson(gson.toJson(objNotification), JsonObject.class));
+        assertEquals(gsonWithNulls.fromJson(objJsonNull, JsonObject.class), gsonWithNulls.fromJson(gsonWithNulls.toJson(objNotification), JsonObject.class));
     }
 
     @Test

